File: framework/pages/automationexercise/AutomationExerciseContactUsPage.ts
import { expect } from '@playwright/test';
import path from 'path';
import { AutomationExerciseBasePage } from './AutomationExerciseBasePage';

/**
 * @pageIdentity AutomationExerciseContactUsPage
 * @urlPattern https://automationexercise.com/contact_us
 */
export class AutomationExerciseContactUsPage extends AutomationExerciseBasePage {
  private contactForm() {
    return this.page.locator('.contact-form');
  }

  public async openFromNav(): Promise<void> {
    await this.dismissCookieConsentIfPresent();
    await this.clickByRole('link', { name: /Contact us/i });
    await this.page.waitForURL(/\/contact_us/, { timeout: 15000 });
  }

  public async assertContactPageLoaded(): Promise<void> {
    await this.assertUrl(/\/contact_us/);
    await expect(this.page.getByText(/GET IN TOUCH/i)).toBeVisible();
  }

  public async fillContactForm(data: {
    name: string;
    email: string;
    subject: string;
    message: string;
  }): Promise<void> {
    await this.dismissCookieConsentIfPresent();
    const form = this.contactForm();
    await form.locator('input[name="name"]').fill(data.name);
    await form.locator('input[name="email"]').fill(data.email);
    await form.locator('input[name="subject"]').fill(data.subject);
    await form.locator('textarea[name="message"]').fill(data.message);
  }

  public async uploadFile(filePath?: string): Promise<void> {
    const resolved =
      filePath ?? path.join(process.cwd(), 'tests', 'fixtures', 'sample.txt');
    await this.contactForm().locator('input[name="upload_file"]').setInputFiles(resolved);
  }

  public async submitAndAcceptAlert(): Promise<void> {
    this.page.once('dialog', (dialog) => dialog.accept());
    await this.contactForm().locator('input[type="submit"][value="Submit"]').click();
  }

  public async assertSubmissionSuccess(): Promise<void> {
    await expect(
      this.page.locator('#contact-page .alert-success').filter({
        hasText: /Success! Your details have been submitted successfully/i,
      })
    ).toBeVisible();
  }

  public async clickHome(): Promise<void> {
    await this.page.locator('#contact-page').getByRole('link', { name: /Home/i }).click();
  }
}

File: framework/pages/automationexercise/automation_exercise_contact_us_page.py
import re
from pathlib import Path

from playwright.sync_api import Page, expect

from framework.config.config_manager import ROOT
from framework.pages.automationexercise.automation_exercise_base_page import (
    AutomationExerciseBasePage,
)


class AutomationExerciseContactUsPage(AutomationExerciseBasePage):
    """@pageIdentity AutomationExerciseContactUsPage
    @urlPattern https://automationexercise.com/contact_us
    """

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    def contact_form(self):
        return self.page.locator(".contact-form")

    def open_from_nav(self) -> None:
        self.dismiss_cookie_consent_if_present()
        self.click_by_role("link", name=re.compile("Contact us", re.I))
        self.page.wait_for_url(re.compile(r"/contact_us"), timeout=15_000)

    def assert_contact_page_loaded(self) -> None:
        self.assert_url(re.compile(r"/contact_us"))
        expect(self.page.get_by_text(re.compile("GET IN TOUCH", re.I))).to_be_visible()

    def fill_contact_form(self, data: dict[str, str]) -> None:
        form = self.contact_form()
        form.locator('input[name="name"]').fill(data["name"])
        form.locator('input[name="email"]').fill(data["email"])
        form.locator('input[name="subject"]').fill(data["subject"])
        form.locator('textarea[name="message"]').fill(data["message"])

    def upload_file(self, file_path: str | Path | None = None) -> None:
        resolved = Path(file_path) if file_path else ROOT / "tests" / "fixtures" / "sample.txt"
        self.contact_form().locator('input[name="upload_file"]').set_input_files(resolved)

    def submit_and_accept_alert(self) -> None:
        self.page.once("dialog", lambda dialog: dialog.accept())
        self.contact_form().locator('input[type="submit"][value="Submit"]').click()

    def assert_submission_success(self) -> None:
        expect(
            self.page.locator("#contact-page .alert-success").filter(
                has_text=re.compile(
                    "Success! Your details have been submitted successfully", re.I
                )
            )
        ).to_be_visible()

    def click_home(self) -> None:
        self.page.locator("#contact-page").get_by_role(
            "link", name=re.compile("Home", re.I)
        ).click()
        try:
            self.page.wait_for_url(re.compile(r"automationexercise\.com/?(?:#.*)?$"), timeout=5_000)
        except Exception:
            self.navigate("https://automationexercise.com/")
